[Torch] Upsampling op support and enable registering a user defined op conversion map

masahi opened the pull request:
This adds support for torch upsample op. It enables running image segmentation models from torchvision (see the test case).

I also added a simple mechanism to register a custom op conversion map, that can be used with the predefined `_convert_map` we already have. The motivation is that torch has too many ops to cover, and it also allows defining custom operators. This makes implementing all op conversion in a single map inside Relay frontend infeasible, and it would be very tedious if we have to modify TVM source each time we need to add a new conversion. Also in torch there are many esoteric operators that doesn't justify for being added to the "official" conversion map in the frontend.

For example of how to use the custom conversion map, see the test case `test_custom_conversion_map()` where I show how to convert torchvision's custom op `torchvision::roi_align` using Relay's `vision.roi_align` op. 

Please review @jwfromm @alexwong @vinx13 @anijain2305 @icemelon9

masahi commented:
@tqchen I get an error from sphinx issues which is not related to this PR, can you take a look? https://ci.tvm.ai/blue/organizations/jenkins/tvm/detail/PR-4961/1/pipeline/150

comaniac commented:
I saw these errors before but have no idea how were they resolved. Could you rebase to upstream master to see if errors persist?

masahi commented:
It is already rebased to the latest commit. The error is still there.
https://github.com/masahi/tvm/commits/torch-more-operators

masahi commented:
@tqchen @comaniac other PRs are blocked by the same CI error.

leandron commented:
On a quick look, *it seems* it is behaving as expected by `task_sphinx_precheck.sh`. It is not a CI unexpected crash.

`task_sphinx_precheck.sh` found warnings during Sphinx execution and it is exiting with status 1 (non-zero), so it fails the Jenkins pipeline.

The relevant lines are:
https://github.com/apache/incubator-tvm/blob/7ccb4363eefa76ae355ee263aa5527d43fb699fb/tests/scripts/task_sphinx_precheck.sh#L43-L46

leandron commented:
Discussion about fixing docs is happening in #4908.

jwfromm commented on python/tvm/relay/frontend/pytorch.py:
```diff
@@ -614,6 +615,51 @@ def _impl(inputs, input_types):
+        cast_func = {
+            6: float,
+        }
+        if inputs[1] in cast_func and not isinstance(data, _expr.Expr):
```
This seems very specific, it looks like it only works for non expression inputs and casts to floats. Is there any reason not to add a `relay.cast` fallback for expression inputs and support other datatypes in the cast_func lookup?

masahi replied:
yeah, the problem is `aten::to(...)` is overloaded for different purpose (transfer to cpu/cuda or simple type cast). See my last commit https://github.com/apache/incubator-tvm/pull/4961/commits/f070d22fdb479c82160cf642e29453472e83c61e
Does this look better?

jwfromm replied:
Ah I didn't know it was overloaded for device copies. This does look much better though, thanks!

comaniac commented:
Yes. It is due to the doc checker introduced yesterday. I'm looking into the problem now.

jwfromm commented on tests/python/frontend/pytorch/test_forward.py:
```diff
@@ -677,9 +666,29 @@ def forward(self, *args):
+    verify_model(Upsample(size=(64, 64), mode="nearest"), inp)
+    verify_model(Upsample(scale=2, mode="nearest"), inp)
+    verify_model(Upsample(size=(64, 64), mode="bilinear", align_corners=True), inp)
+    verify_model(Upsample(scale=2, mode="bilinear", align_corners=True), inp)
```
Probably worth adding a bilinear test with non integer scaling, i.e. `size = (50, 50)`.

masahi replied:
added

jwfromm commented:
Left minor comments but otherwise looks good. Having a custom convert map is a neat idea that I'd like to see added to other frontends.

tqchen commented:
docs error should be fixed by https://github.com/apache/incubator-tvm/pull/4967

alexwong commented on tests/python/frontend/pytorch/test_forward.py:
```diff
@@ -23,6 +23,7 @@
 from scipy.stats import t as tdistr
 import numpy as np
 import torch
+from torch import nn
```
Nitpick: There's places where we use torch.nn and others where we use nn, can we change the former to the latter and remove the below import as well.

masahi replied:
Removed `from torch import nn` and updated to use `torch.nn`

masahi commented:
@alexwong @jwfromm I removed the dependency on `packaging` and dropped support for PT 1.2 and  older per discussion https://discuss.tvm.ai/t/pytorch-frontend-no-module-named-packaging/5829.

masahi commented:
@vinx13 @anijain2305 @icemelon9 comments are addressed and tests passed, I think it is ready for merge. A custom convert map could be of interest to other frontends.

anijain2305 commented on tests/python/frontend/pytorch/test_forward.py:
```diff
@@ -730,6 +752,55 @@ def test_vgg11_bn():
+
+    for model in [fcn, deeplab]:
+        verify_model(SegmentationModelWrapper(model.eval()), inp,
+                     ctx_list=[("cuda", tvm.gpu(0))])  # dilated covolution not supported on x86
```
Maybe adding the issue number is helpful here?

masahi commented:
Thanks @jwfromm @alexwong @anijain2305
